jerry-core/ecma/operations/ecma-function-object.c

```diff
@@ -422,27 +463,32 @@ ecma_op_function_call (ecma_object_t *func_obj_p, /**< Function object */
                 && !ecma_is_lexical_environment (func_obj_p));
   JERRY_ASSERT (ecma_op_is_callable (ecma_make_object_value (func_obj_p)));
 
-  ecma_value_t ret_value = ECMA_VALUE_EMPTY;
-
-  if (ecma_get_object_type (func_obj_p) == ECMA_OBJECT_TYPE_FUNCTION)
+  while (true)
   {
-    if (JERRY_UNLIKELY (ecma_get_object_is_builtin (func_obj_p)))
-    {
-      ret_value = ecma_builtin_dispatch_call (func_obj_p,
-                                              this_arg_value,
-                                              arguments_list_p,
-                                              arguments_list_len);
-    }
-    else
+    ecma_object_type_t func_type = ecma_get_object_type (func_obj_p);
+
+    JERRY_ASSERT (func_type == ECMA_OBJECT_TYPE_FUNCTION
+                  || !ecma_op_function_has_construct_flag (arguments_list_p));
+
+    if (func_type == ECMA_OBJECT_TYPE_FUNCTION)
     {
+      if (JERRY_UNLIKELY (ecma_get_object_is_builtin (func_obj_p)))
+      {
+        return ecma_builtin_dispatch_call (func_obj_p,
```

Comment:
Built-in function objects **_should  never_** have a `construct_flag`, but it would be nice to see a `JERRY_ASSERT (!ecma_op_function_has_construct_flag (arguments_list_p))` before returning with `ecma_builtin_dispatch_call`.

This case currently seems impossible but the assertion could prevent the wrong usage of the `ecma_op_function_set_construct_flag` in the future.